### PR TITLE
HAWQ-362. Master RM should Recalculate Segment's Allocation Status if Segment's RM is restart

### PR DESCRIPTION
**Pull request by linwen**
Please review, Thanks!

### CODE REVIEW

File: src/backend/resourcemanager/resourcepool.c [hunk 1]
```diff
@@ -893,6 +893,28 @@ int addHAWQSegWithSegStat(SegStat segstat, bool *capstatchanged)
...
+			 * This segment's RM process has restarted,
+			 * we should clean up old status, so mark it down.
+			 */
+			if (oldStatus == RESOURCE_SEG_STATUS_AVAILABLE && !statusChanged)
+			{
+				segstat->FTSAvailable = RESOURCE_SEG_STATUS_UNAVAILABLE;
+				statusChanged = true;
+			}
+			segresource->Stat->RMStartTimestamp = segstat->RMStartTimestamp;
+			elog(LOG, "Master RM finds segment:%s 's RM process has restarted. "
```
Comment (jiny2) [inline src/backend/resourcemanager/resourcepool.c]: Suggest to report as a WARNING log.

**Comment by jiny2** [comment]
The code change is ok to me. +1

File: src/backend/resourcemanager/resourcepool.c [hunk 2]
```diff
@@ -893,6 +893,28 @@ int addHAWQSegWithSegStat(SegStat segstat, bool *capstatchanged)
 	else {
 		segresource = getSegResource(segid);
 		Assert(segresource != NULL);
+		uint8_t oldStatus = segresource->Stat->FTSAvailable;
+		bool statusChanged = oldStatus != segstat->FTSAvailable;
+
+		/* Check if RM process is restarted in this segment */
+		if (segresource->Stat->RMStartTimestamp != segstat->RMStartTimestamp)
```
Comment (huor) [inline src/backend/resourcemanager/resourcepool.c]: A comment here to illustrate the idea would be better. It is also good to mention the risk that below three condition hold:
1. rmseg reset
2. system time reset on segment node
3. segresource->Stat->RMStartTimestamp == segstat->RMStartTimestamp

**Comment by huor** [comment]
The overall fix looks good. +1

**Pull request closed without merging**